text/0000-sandbox-environment.md

```diff
@@ -0,0 +1,193 @@
+distinct from the actual process environment. By default they are the same, but the additions of the `--env-whitelist`,
+`--env-blacklist` and `--env-set` options allow the logical environment to be tailored as desired.
+
+## Processing of the options
+
+The `--env-` options are processed in the order they appear on the command-line, left to right. The logical environment is
+initialized from the process environment. Then each each `--env` option is processed in turn, as it appears, to update the logical
+environment. Specifically:
+
+- `--env-whitelist REGEX` - Any name which doesn't match the REGEX is removed from the logical environment,
+  as if it had never been set. This is symmetric with `--env-blacklist`.
+- `--env-blacklist REGEX` - Any name which does match the REGEX is removed from the logical environment, as if it had never
+  been set. This is symmetric with `--env-whitelist`.
+- `--env-set VAR=VALUE` - Set a logical environment variable with the given value. This either sets a new variable, or
+  overrides an existing variable's value.
+
+Note that `--env-whitelist` and `--env-blacklist` affect variables set with previous `--env-set` options, possibly removing them.
+
+If there are no `--env-` options then the logical environment is left in its initial state, which is identical to the process
+environment.
+
+`rustc` will only accept UTF-8 encoded command-line options, which affects all these options. TBD: What happens if the 
+process environment has non-UTF-8 names or values?
+
+## Compile-time behaviour
+
+The `env!()` and `option_env!()` macros strictly act on the logical environment variables with no reference
+to the process environment.
+
+Note that this can't affect other environment accesses. For example, if a procedural macro uses the `std::env::var()` function
+it will access the process environment. Any process that happens to be invoked by `rustc` would still see the original
+process environment, not the logical environment.
+
+(TBD: describe internal API for accessing the logical environment.)
+
+## Cargo
+
+This has no direct effect on Cargo - it can completely ignore these options and the overall behaviour would be unchanged. However,
+it's easy to imagine a corresponding RFC for Cargo where it does more explicitly control the logical environment. For example,
+it could constrain the accessible variables to:
+1. ones that Cargo itself sets
+2. ones that the build script sets via `rustc-env`
+3. ones that the build script notes as `rerun-if-env-changed`
+4. explicitly listed in the `Cargo.toml`
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The primary cost is additional complexity in invoking `rustc` itself, and additional complexity in documenting
+`env!`/`option_env!`. Procedual macros would need to be changed to access the logical environment, either by
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
```

Comment:
[`env(1)`](https://linux.die.net/man/1/env) seems like a plausible alternative to me.

Reply:
Sure, I'll add it, and explicitly using the shell.